src/commands.rs

```diff
@@ -4150,6 +4151,130 @@ fn do_git_clone(
+    mut_repo: &mut MutableRepo,
+    git_repo: &git2::Repository,
+    remote_name: &str,
+) -> Result<Option<String>, GitFetchError> {
+    let mut callback = None;
+    if atty::is(atty::Stream::Stdout) {
+        let mut rate = RateEstimate::new();
+
+        struct Guard<'a> {
+            ui: &'a mut Ui,
+            printed: bool,
+        }
+        impl Drop for Guard<'_> {
+            fn drop(&mut self) {
+                if self.printed {
+                    let _ = writeln!(self.ui);
+                }
+            }
+        }
+        let mut guard = Guard { ui, printed: false };
+
+        callback = Some(move |progress: &git::Progress| {
+            const HIDE_CURSOR: &str = "\x1b[25l";
+            const MOVE_TO_START: &str = "\x1b[G";
+            const SHOW_CURSOR: &str = "\x1b[25h";
+            const CLEAR_TRAILING: &str = "\x1b[K";
+            let _ = write!(
+                guard.ui,
+                "{}{}{: >3.0}%",
+                HIDE_CURSOR,
+                MOVE_TO_START,
+                100.0 * progress.overall
+            );
+            if let Some(estimate) = progress
+                .bytes_downloaded
+                .and_then(|x| rate.update(Instant::now(), x))
+            {
+                let (scaled, prefix) = binary_prefix(estimate);
+                let _ = write!(
+                    guard.ui,
+                    "at {: >5.1} {}B/s{}{}",
+                    scaled, prefix, CLEAR_TRAILING, SHOW_CURSOR,
+                );
+                guard.printed = true;
+            }
+            let _ = write!(guard.ui, "{}{}", CLEAR_TRAILING, SHOW_CURSOR);
+        });
+    }
+    let result = git::fetch(
+        mut_repo,
+        git_repo,
+        remote_name,
+        callback.as_mut().map(|x| x as _),
+    );
+    result
+}
+
+/// Find the smallest binary prefix with which the whole part of `x` is at most
+/// three digits, and return the scaled `x` and that prefix.
+fn binary_prefix(x: f32) -> (f32, &'static str) {
+    const TABLE: [&str; 9] = ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi", "Yi"];
+
+    let mut i = (x.abs().log(1024.0) as usize).min(TABLE.len() - 1);
+    let mut scaled = x / 1024.0f32.powi(i as i32);
+    if scaled > 1000.0 {
+        i = (i + 1).min(TABLE.len() - 1);
+        scaled /= 1024.0;
+    }
+    (scaled, TABLE[i])
+}
+
+struct RateEstimate {
+    state: Option<RateEstimateState>,
+}
+
+struct RateEstimateState {
+    total: u64,
+    avg_rate: Option<f32>,
+    last_sample: Instant,
+}
+
+impl RateEstimate {
+    fn new() -> Self {
+        RateEstimate { state: None }
+    }
+
+    /// Compute smoothed rate from an update
+    fn update(&mut self, now: Instant, total: u64) -> Option<f32> {
+        let state = match self.state {
+            None => {
+                self.state = Some(RateEstimateState {
+                    total,
+                    avg_rate: None,
+                    last_sample: now,
+                });
+                return None;
```

Comment:
It looks like we always return `self.state.avg_rate` (unwrapped), which makes sense. Can we clarify that in the code?

Maybe something like this:

```
match self.state {
  None => {
    self.state = Some(RateEstimateState { ... });
  }
  Some(state) => {
    // Move most of the calculation into a new RateEstimateState::update()
    state.update(now, total);
  }
}
let state = self.state.as_ref().unwrap();
state.avg_rate.clone()
```

I'm not sure that will be clearer, so you decide.

Reply:
Split the difference using `insert` and returning from the inner method, which is a clear improvement. I prefer not to rely on `unwrap` when static checking is feasible.